### PR TITLE
Get all unit data from graph database

### PR DESCRIPTION
**Pull request by cmutel**

### CODE REVIEW

File: units/routes.py [hunk 1]
```diff
@@ -29,27 +28,53 @@ def search(
...
 
-    Returns:
-        FullConceptSchemeResponse: The concept scheme with member
-            concepts and collections.
     """
-    return controller.get_concept_scheme(concept_scheme_iri, lang=lang)
+    if lang and not len(lang) == 2:
+        raise HTTPException(
+            status_code=500, detail="Lang string must be exactly two letters long"
+        )
```
Comment (jsvgoncalves) [inline units/routes.py:60]: Move input validation to pydantic.

File: units/routes.py [hunk 2]
```diff
@@ -29,27 +28,53 @@ def search(
...
+
+    qk = get_qk_for_iri(iri)
+    if not qk:
+        raise HTTPException(status_code=404, detail="Unit IRI not found")
```
Comment (jsvgoncalves) [inline units/routes.py:64]: Raise in the `get_or_raise()`. (this can be done later for sure)

File: units/settings.py [hunk 3]
```diff
@@ -9,11 +10,10 @@
...
 
-    #API_KEY: SecretStr
-    #DATABASE_URL: SecretStr
-    SENTRY_DSN: SecretStr = SecretStr("")
-
-    HOST_IP: str = "0.0.0.0"
+    UNITS_SENTRY_DSN: Optional[SecretStr] = SecretStr("")
+    UNITS_SPARQL_URL: Optional[str] = "https://fuseki.d-d-s.ch/skosmos/query"
+    UNITS_VOCAB_PREFIX: Optional[str] = "https://vocab.sentier.dev/"
+    UNITS_HOST_IP: str = "0.0.0.0"
```
Comment (jsvgoncalves) [inline units/settings.py:16]: Would be better to use [`env_prefix="UNITS_"`][prefix] below.

[prefix]: https://docs.pydantic.dev/latest/concepts/pydantic_settings/#validation-of-default-values

File: units/concepts.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+def language_filter(o: dict, lang: str) -> dict:
+    """Check if given element has the correct language, if one is given"""
+    if "xml:lang" in o and lang not in o["xml:lang"].lower():
+        return False
+    return True
+
+
+def get_qk_for_iri(iri: str) -> str | None:
+    """Get the QUDT quantity key for a given unit IRI."""
+    settings = get_settings()
```
Comment (jsvgoncalves) [inline units/concepts.py:34]: This could initialise at the module level instead, just after the logger.

File: units/concepts.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+
+def get_all_data_for_qk_iri(
+    iri: str,
+    lang: str | None = None,
+    remove_namespaces: bool = True,
+    graph_namespaces: list[str] = ["qudt", "simapro"],
+) -> dict:
+    """Get all data for a given quantity kind IRI."""
+    settings = get_settings()
```
Comment (jsvgoncalves) [inline units/concepts.py:63]: Same as above.

File: units/concepts.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    logger.debug("Using sparql endpoint url %s", settings.UNITS_SPARQL_URL)
+
+    QUERY = f"""
+PREFIX qudt: <http://qudt.org/schema/qudt/>
+
+JSON {{
+    "this is ignored anyway": ?qk
+}}
+FROM <{settings.UNITS_VOCAB_PREFIX}qudt/>
+FROM <{settings.UNITS_VOCAB_PREFIX}simapro/>
```
Comment (jsvgoncalves) [inline units/concepts.py:44]: Is this going to expand over time? I guess it could later be dynamic?

File: units/concepts.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+WHERE {{
+    <{iri}> qudt:hasQuantityKind ?qk .
+}}
+    """
+
+    logger.debug("Executing query %s", QUERY)
+    response = httpx.post(settings.UNITS_SPARQL_URL, data={"query": QUERY}).json()
+    logger.info(f"Retrieved {len(response)} quantity kind results for iri {iri}")
```
Comment (jsvgoncalves) [inline units/concepts.py:52]: A `raise_for_status` wouldn't hurt.

File: units/concepts.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    lang: str | None = None,
+    remove_namespaces: bool = True,
+    graph_namespaces: list[str] = ["qudt", "simapro"],
+) -> dict:
+    """Get all data for a given quantity kind IRI."""
+    settings = get_settings()
+    logger.debug("Using sparql endpoint url %s", settings.UNITS_SPARQL_URL)
+
+    if lang and not len(lang) == 2:
+        raise ValueError("Language code must be exactly two letters long")
```
Comment (jsvgoncalves) [inline units/concepts.py:67]: Here we could reuse the same pydantic as for the route and avoid the double validation.

File: units/concepts.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            len(response),
+            iri,
+            settings.UNITS_VOCAB_PREFIX + graph_namespace,
+        )
+        results.extend(response)
+
+    if remove_namespaces:
+        formatter = remove_graph_namespaces
+    else:
+        formatter = lambda x: x
```
Comment (jsvgoncalves) [inline units/concepts.py:98]: Pass the formatter as arg?

**Review by jsvgoncalves — COMMENTED** [review]
See below, we can merge and fix later (just don't leave issues unresolved).

File: units/routes.py (hunk 1 above)
Comment (cmutel) [reply units/routes.py:60]: Removed this as it doesn't actually work for our use case

File: units/concepts.py (hunk 6 above)
Comment (cmutel) [reply units/concepts.py:44]: Rather the opposite - see [comment ](https://github.com/Depart-de-Sentier/brightcon-2024-material/issues/1#issuecomment-2370351664)

File: units/concepts.py (hunk 8 above)
Comment (cmutel) [reply units/concepts.py:67]: Removed this code

File: units/routes.py (hunk 2 above)
Comment (cmutel) [reply units/routes.py:64]: Raised a KeyError in origin function

File: units/concepts.py (hunk 9 above)
Comment (cmutel) [reply units/concepts.py:98]: Refactoring this out in any case :)

**Pull request merged**